Re-use async to support WANT_WRITE while sending fragments

julek-wolfssl opened the pull request:
- Async I/O can be turned off with `WOLFSSL_NO_ASYNC_IO`
- `WOLFSSL_ASYNC_IO` functionality enabled in `SendCertificateVerify()` and `SendServerKeyExchange()` to allow safe re-entry into SendHandshakeMsg()
- Testing size of structs is refactored in `WOLFSSL_ASSERT_SIZEOF_GE()`

The error can be replicated with:

```
./configure CPPFLAGS='-DWOLFSSL_STATIC_RSA -DWOLFSSL_STATIC_DH'
./examples/server/server -v 3 -6
./examples/client/client -v 3 -l TLS_DHE_RSA_WITH_CHACHA20_POLY1305_SHA256 -6
```

Cipher selection is important to force a cert verify to be sent.

julek-wolfssl committed:
Re-use async to support WANT_WRITE while sending fragments

- Async I/O can be turned off with WOLFSSL_NO_ASYNC_IO
- WOLFSSL_ASYNC_IO functionality enabled in SendCertificateVerify() and SendServerKeyExchange() to allow safe re-entry into SendHandshakeMsg()
- Testing size of structs is refactored int WOLFSSL_ASSERT_SIZEOF_GE()

julek-wolfssl committed:
Use WOLFSSL_ASYNC_IO for WOLFSSL_NONBLOCK_OCSP

- Enable ssl->async to store function arguments for non-blocking OCSP
- Remove ssl->nonblockarg

dgarske reviewed (changes requested):
This is great work here, but we need to find a solution that doesn't increase the WOLFSSL object size by default. The PR currently enables `WOLFSSL_ASYNC_IO` for most use-cases, which adds over 100 bytes to the WOLFSSL object. I'd prefer we find a way to use the heap for these edge cases to retain the stack variables (only when needed for re-entry). If you are not able to use the the async code for this then it could be that your original solution in PR #5134 is better. Or maybe try and leverage the OCSP non-blocking logic?

rliebscher commented:
I tried the patch with 5.3.0 and did not get any of the previous problems of the bug, so it seems to fix #5108.

julek-wolfssl commented:
@rliebscher Thank you for being diligent and checking!

julek-wolfssl committed:
Allocate ssl->async seperately to the SSL object

rizlik commented on src/internal.c:
```diff
@@ -26474,14 +26512,23 @@ int SendCertificateVerify(WOLFSSL* ssl)
-    #ifdef WOLFSSL_ASYNC_CRYPT
-        ssl->async.freeArgs = FreeScvArgs;
+    #ifdef WOLFSSL_ASYNC_IO
+        ssl->async->freeArgs = FreeScvArgs;
```
Also here, I think in the cleanup of the function:
```    /* Final cleanup */
    FreeScvArgs(ssl, args);
    FreeKeyExchange(ssl);
```
FreeScvArgs is invoked twice.

julek-wolfssl replied:
Fixed.

rizlik commented on src/sniffer.c:
```diff
@@ -3092,6 +3100,10 @@ static int SetupKeys(const byte* input, int* sslBytes, SnifferSession* session,
     FreeSetupKeysArgs(ssl, args);
 
+#ifdef WOLFSSL_ASYNC_CRYPT
+    FreeAsyncCtx(WOLFSSL* ssl);
```
FreeSetupKeysArgs invoked twice.

rizlik reviewed (changes requested):
I like the idea of re-using crypto async for async i/o. I think there is also room for better integration in wolfSSL_connect/wolfSSL_accept, so we don't have to "guess" whether we should send or advance the state. 
On the dynamic allocation of the `ssl->async` object, I would also consider the allocation of the args field only, as it will simplify some memory management.

julek-wolfssl commented on src/internal.c:
```diff
@@ -11928,15 +11944,13 @@ int ProcessPeerCerts(WOLFSSL* ssl, byte* input, word32* inOutIdx,
+#endif
+#ifdef WOLFSSL_NONBLOCK_OCSP
+    if (ssl->error == OCSP_WANT_READ) {
+        /* Re-entry after non-blocking OCSP */
```
They only emit a warning if the `{ }` braces are omitted.

julek-wolfssl commented on src/internal.c:
```diff
@@ -6980,15 +6996,9 @@ void FreeKeyExchange(WOLFSSL* ssl)
-#ifndef WOLFSSL_NO_TLS12
-    FreeBuildMsgArgs(ssl, &ssl->async.buildArgs);
-#endif
+    FreeAsyncCtx(ssl);
```
Moved `FreeAsyncCtx` outside of `FreeKeyExchange`.

dgarske reviewed (changes requested):
The async PR is looking good. It all tests out correctly with the QAT hardware. I will need to compare performance impact still. In order to avoid always allocating / freeing heap memory during the handshake I would prefer to allocate when args is first needed and keep until the handshake is done. Free during the `FreeHandshakeResources`.

Note: I had to make this patch to async.c:
```
diff --git a/wolfcrypt/src/async.c b/wolfcrypt/src/async.c
index aea068e..132a96e 100644
--- a/wolfcrypt/src/async.c
+++ b/wolfcrypt/src/async.c
@@ -44,7 +44,7 @@ static WC_ASYNC_DEV* wolfAsync_GetDev(WOLF_EVENT* event)
             case WOLF_EVENT_TYPE_ASYNC_WOLFSSL:
             {
                 WOLFSSL* ssl = (WOLFSSL*)event->context;
-                dev = ssl->async.dev;
+                dev = ssl->asyncDev;
                 break;
             }
```

julek-wolfssl commented:
@dgarske `ssl->async` is now allocated when necessary but only free'd at the end of the handshake (unless doing async crypto when it is only free'd with the `WOLFSSL` object). Implemented by adding a `freeAsync` parameter to `FreeAsyncCtx`.

julek-wolfssl commented:
You can observe the fewer allocations with valgrind's allocation count.

dgarske reviewed (changes requested):
I have compared code size (only ~200 byte difference), memory use and benchmarks. All look great! I also tested the OCSP non-blocking to confirm it still works.

My only suggestion at this point is to add a test for the original use-case that caused this work:

```
./configure CPPFLAGS='-DWOLFSSL_STATIC_RSA -DWOLFSSL_STATIC_DH'
./examples/server/server -v 3 -6
./examples/client/client -v 3 -l TLS_DHE_RSA_WITH_CHACHA20_POLY1305_SHA256 -6
```

dgarske commented on wolfssl/wolfcrypt/settings.h:
```diff
@@ -2671,6 +2671,16 @@ extern void uITRON4_free(void *p) ;
     #define NO_RC4
 #endif
 
+#if !defined(WOLFSSL_NO_ASYNC_IO) || defined(WOLFSSL_ASYNC_CRYPT) || \
```
Consider adding `WOLFSSL_NO_ASYNC_IO` to documentation... and to the configure `--enable-lowresource`.

julek-wolfssl replied:
Added to our macro docs and added to the low resource build.

dgarske commented on wolfssl/internal.h:
```diff
@@ -4283,10 +4292,13 @@ struct WOLFSSL {
-    struct WOLFSSL_ASYNC async;
-#elif defined(WOLFSSL_NONBLOCK_OCSP)
-    void*           nonblockarg;        /* dynamic arg for handling non-block resume */
+    WC_ASYNC_DEV* asyncDev;
```
Is `WC_ASYNC_DEV* asyncDev;` still needed/used?

julek-wolfssl replied:
Yes, it is needed for async crypto. It also needs to be separate because it is not only used within the handshake machine context.

julek-wolfssl commented:
Would be good to test write blocking with SCR as well.

julek-wolfssl committed:
Add WOLFSSL_NO_ASYNC_IO to low resource build

julek-wolfssl committed:
Fix curve group matching on secure renegotiation

julek-wolfssl committed:
Add testing/docs for blocking write

- Fix case where message grouping can make CheckAvailableSize return a WANT_WRITE
- CheckAvailableSize in tls13.c will not return a WANT_WRITE since it only does so for DTLS <=1.2

dgarske reviewed (approved):
Changes all look good! Tested with Intel QuickAssist, simulator, OCSP stapling non-blocking and with `WOLFSSL_NO_ASYNC_IO`.

julek-wolfssl committed:
Simulate WANT_WRITE only with async I/O support

julek-wolfssl commented:
All checks now passing.

rizlik commented on src/internal.c:
```diff
@@ -12007,15 +12032,13 @@ int ProcessPeerCerts(WOLFSSL* ssl, byte* input, word32* inOutIdx,
     }
-    if (ssl->nonblockarg == NULL) /* new args */
+    else
+#endif
```
[can't put the comment on the right lines because they aren't part of the diff. lines 12045
Is the 
```
if (args == NULL) {
        ERROR_OUT(MEMORY_E, exit_ppc);
}
```
check outside of the else branch of line 12040 on purpose?
I don't think it hurts, but the indentation is misleading.

julek-wolfssl replied:
The `else` applies to the `/* Reset state */` block. The block you cited is in the `elif` when building for small stack.

rizlik commented on src/internal.c:
```diff
@@ -13361,9 +13384,6 @@ int ProcessPeerCerts(WOLFSSL* ssl, byte* input, word32* inOutIdx,
 #endif /* WOLFSSL_ASYNC_CRYPT || WOLFSSL_NONBLOCK_OCSP || WOLFSSL_SMALL_STACK */
 
 #if defined(WOLFSSL_ASYNC_CRYPT)
-#elif defined(WOLFSSL_NONBLOCK_OCSP)
```
NIT: change
```
#if defined(WOLFSSL_ASYNC_CRYPT)
#elif defined(WOLFSSL_SMALL_STACK)
    XFREE(args, ssl->heap, DYNAMIC_TYPE_TMP_BUFFER);
#endif
```
in
```
#if !defined(WOLFSSL_ASYNC_CRYPT) && defined((WOLFSSL_SMALL_STACK)
    XFREE(args, ssl->heap, DYNAMIC_TYPE_TMP_BUFFER);
#endif
```

julek-wolfssl replied:
Fixing.

rizlik commented on src/internal.c:
```diff
@@ -20534,8 +20608,16 @@ int SendAlert(WOLFSSL* ssl, int severity, int type)
     outputSz = ALERT_SIZE + MAX_MSG_EXTRA + dtlsExtra;
-    if ((ret = CheckAvailableSize(ssl, outputSz)) != 0)
-        return ret;
+    if ((ret = CheckAvailableSize(ssl, outputSz)) != 0) {
```
should we set `ssl->options.buildingMsg = 1` before exit in case of `ret == WANT_WRITE && severity != alert_fatal`?

julek-wolfssl replied:
I think that the higher layer should have set that don't you think?

julek-wolfssl replied:
What do you think about ignoring the alert if it is non fatal and we get a `WANT_WRITE` here?

rizlik replied:
> I think that the higher layer should have set that don't you think?

good point. I think I'm ok with the actual solution then.

> What do you think about ignoring the alert if it is non-fatal and we get a WANT_WRITE here?

I don't know how much a non-fatal alert is important for the protocol.

julek-wolfssl replied:
> I don't know how much a non-fatal alert is important for the protocol.

That's a good point. I'll think about how to best solve this issue.

julek-wolfssl replied:
Do you think the pending alert solution will solve the scenario you described?

rizlik reviewed (dismissed):
only minor things and a question

julek-wolfssl committed:
Combine preproc check into one statement

julek-wolfssl committed:
Save pending alerts when using async io

- Don't overwrite ssl->error
- Clear the error in ssl->error because the return of SendBuffered is now stored in ret instead

dgarske reviewed (changes requested):
Code changes all look great. Most tests pass, but there are some edge case with QAT hardware occurring.

I tracked it down to the `DoHandShakeMsg` returning `WC_PENDING_E`, but it not being captured in the `ssl->error`. It looks like some of the `ret` and `ssl->error` refactoring broke it.

```
./configure --enable-all --enable-asynccrypt --with-intelqa=../QAT1.8 --enable-debug --disable-shared && make
sudo tests/unit.test 
...
processing server hello done
wolfSSL Leaving DoHandShakeMsgType(), return 0
wolfSSL Leaving DoHandShakeMsg(), return 0
connect state: HELLO_AGAIN
connect state: HELLO_AGAIN_REPLY
connect state: FIRST_REPLY_DONE
wolfSSL Entering SendCertificate
growing output buffer
wolfSSL Leaving SendCertificate, return 0
sent: certificate
connect state: FIRST_REPLY_FIRST
wolfSSL Entering SendClientKeyExchange
wolfSSL Leaving wolfSSL_AsyncPop, return -109
wolfSSL Entering EccMakeKey
wolfSSL Leaving wolfSSL_AsyncInit, return 0
wolfSSL Leaving wolfSSL_AsyncPush, return -108
wolfSSL Leaving EccMakeKey, return -108
wolfSSL Leaving SendClientKeyExchange, return -108
wolfSSL Entering SSL_get_error
wolfSSL Leaving SSL_get_error, return 0
   wolfSSL_get_client_CA_list() with handshake:ret -1, err 0
ERROR - tests/api.c line 33919 failed with:
    expected: ret == WOLFSSL_SUCCESS
    result:   -1 != 1
Aborted
```

Also another case:

```
processing server key exchange
wolfSSL Entering DoServerKeyExchange
wolfSSL Leaving wolfSSL_AsyncPop, return -109
wolfSSL Entering RsaVerify
wolfSSL Leaving wolfSSL_AsyncInit, return 0
wolfSSL Leaving wolfSSL_AsyncPush, return -108
wolfSSL Leaving RsaVerify, return -108
wolfSSL Leaving DoServerKeyExchange, return -108
wolfSSL Leaving DoHandShakeMsgType(), return -108
wolfSSL Leaving DoHandShakeMsg(), return -108
wolfSSL Entering SSL_get_error
wolfSSL Leaving SSL_get_error, return 0
   wolfSSL_get_client_CA_list() with handshake:ret -1, err 0

ERROR - tests/api.c line 33920 failed with:
    expected: ret == WOLFSSL_SUCCESS
    result:   -1 != 1

```

julek-wolfssl committed:
Store RetrySendAlert error in ssl->error

julek-wolfssl commented:
I wasn't able to reproduce your error with your patch (note that you are not setting the device ID and getting the error from `ssl` (wrong `WOLFSSL` object) not `ssl_client`). I looked over the last commit and this is the only area where I found that `ssl->error` wasn't being set.
```
diff --git a/src/internal.c b/src/internal.c
index 897babad7..f78299161 100644
--- a/src/internal.c
+++ b/src/internal.c
@@ -20294,8 +20294,10 @@ int SendData(WOLFSSL* ssl, const void* data, int sz)
     }
 
     ret = RetrySendAlert(ssl);
-    if (ret != 0)
-        return ret;
+    if (ret != 0) {
+        ssl->error = ret;
+        return WOLFSSL_FATAL_ERROR;
+    }
 
     for (;;) {
         byte* out;
```

julek-wolfssl committed:
SendAlert: clear output buffer to try and send the alert now

dgarske reviewed (approved):
This PR is approved. Excellent work on this Juliusz.

The QAT issue was not related to this PR and fixed in https://github.com/wolfSSL/wolfssl/pull/5198
